### PR TITLE
Compatibility for the "Use Unicode UTF-8 for worldwide language support" feature

### PR DESCRIPTION
**Pull request by AAClause**
### Link to issue number:
Fixes #8599

### Summary of the issue:
Windows 10 1803 (April 2018 Update) introduced a new feature, "Use Unicode UTF-8 for worldwide language support". When this feature is enabled, add-ons fail to install.

### Description of how this pull request fixes the issue:
The solution consists to alias cp65001 to UTF-8.

### Testing performed:
Tested with the feature enabled, NVDA portable and installed. I tried to install several add-ons with success.

### Known issues with pull request:
None

### Change log entry:
Bug fixes:
* In Windows 10 1803 and later, it is now possible to install add-ons if "Use Unicode UTF-8 for worldwide language support" feature is enabled. (#8599)

### CODE REVIEW

**Review by josephsl — COMMENTED** [review]
Hi,

Is cp65001 the only new code page? I'm sure there might be additional code pages that should be "labeled" as UTF-8.

Thanks.

**Comment by Andre9642** [comment]
Hi,
According [this](https://docs.microsoft.com/en-us/windows/desktop/intl/code-page-identifiers), yes.
Apparently, Python 3.3 fixes this bug, see https://bugs.python.org/issue13216

**Comment by josephsl** [comment]
Hi, in this case, can you include this in your code comment so we can come back to it when we work on #7105? Thanks.

**Comment by zstanecic** [comment]
hi to all,

is this going to be merged in master, or we will wait for #7104


07.08.2018 15:52, Joseph Lee пишет:

**Comment by josephsl** [comment]
Hi, once we know for sure the CP65001 is the only format that needs to be added, then I think it should be part of 2018.3. Thanks.

**Comment by zstanecic** [comment]
btw,

this also affects your add-ons i.e add-on updater cannot install 
add-ons, and

addons cannot be installed when the operating system is set to use this 
corresponding locale

07.08.2018 16:04, Joseph Lee пишет:

**Comment by josephsl** [comment]
Hi, I’ll mitigate this from my add-on in the next release. Thanks.

**Comment by josephsl** [comment]
Hi, note that this is something that must be done from Core. As I said, I’ll mitigate this from my add-on (Add-on Updater) in the next release, as that ad-on uses code that’s destined for NVDA Core later. Thanks..

File: source/gui/__init__.py [hunk 1]
```diff
@@ -35,6 +35,10 @@
 import guiHelper
 import winVersion
 
+# Temporary: #8599: add cp65001 codec
+#            #7105: upgrading to python 3 should fix this issue. See https://bugs.python.org/issue13216
+codecs.register(lambda name: codecs.lookup('utf-8') if name == 'cp65001' else None)
```
Comment (leonardder) [inline source/gui/__init__.py]: Is the `codecs.register(lambda name: None)` gracefully handled by codecs? The lambda is expected to return a tuple of functions (encoder, decoder, stream_reader, stream_writer) (or a CodecInfo object), according to the docs. I've only checked python's built-in help, though.

**Comment by Andre9642** [comment]
> Is the codecs.register(lambda name: None) gracefully handled by codecs?

According to [the documentation](https://docs.python.org/2/library/codecs.html#codecs.register):

> <dl><dt>codecs.register(search_function)<//dt>
> <dd>Register a codec search function. Search functions are expected to take one argument, the encoding name in all lower case letters, and return a CodecInfo object having the following attributes:  
> [...]
>
> In case a search function cannot find a given encoding, it should return None.
>
> Encodings are first looked up in the registry’s cache. If not found, the list of registered search functions is scanned. If no CodecInfo object is found, a LookupError is raised. Otherwise, the CodecInfo object is stored in the cache and returned to the caller.
> [...]</dd></dl>

**Review by LeonarddeR — DISMISSED** [review]
Great, thanks for clarifying.

**Comment by leonardder** [comment]
[**@josephsl**](https://github.com/josephsl) commented on [7 aug. 2018 16:04 CEST](https://github.com/nvaccess/nvda/pull/8602#issuecomment-411067809 "2018-08-07T14:04:16Z - Replied by Github Reply Comments"):
> Hi, once we know for sure the CP65001 is the only format that needs to be added, then I think it should be part of 2018.3. Thanks.

I tend to agree. Note that this will require a change of the base branch.

**Comment by michaelDCurran** [comment]
I have switched the base branch to beta and rebased, fixing up changes.t2t.

**Review by michaelDCurran — APPROVED** [review]

**Pull request merged**